File: programmers/lv1/신규아이디추천.py
def solution(new_id):
    answer = new_id.lower()
    answer = answer.replace("~","").replace("#","").replace("@","").replace("$","").replace("%","").replace("^","").replace("&","").replace("*","").replace("(","").replace(")","").replace("=","").replace("+","").replace("[","").replace("{","").replace("]","").replace("}","").replace(":","").replace("?","").replace("!","").replace(",","").replace("<","").replace(">","").replace("/","")
    newWord = ''
    if len(answer) == 1:
        newWord = answer
    else:
        for i in range(len(answer) - 1):
            if answer[i] == '.' and answer[i+1] == '.':
                continue
            newWord += answer[i]
            if i == len(answer) - 2:
                newWord += answer[i+1]

    newWord = newWord.strip(".")

    if len(newWord) == 0:
        newWord = 'a'

    if len(newWord) >= 16:
        newWord = newWord[:15]

    newWord = newWord.strip(".")

    if len(newWord) <= 2:
        while(len(newWord) < 3):
            newWord += newWord[-1]
    return newWord

solution("b")

# best
import re

def solution(new_id):
    st = new_id
    st = st.lower()
    st = re.sub('[^a-z0-9\-_.]', '', st)
    st = re.sub('\.+', '.', st)
    st = re.sub('^[.]|[.]$', '', st)
    st = 'a' if len(st) == 0 else st[:15]
    st = re.sub('^[.]|[.]$', '', st)
    st = st if len(st) > 2 else st + "".join([st[-1] for i in range(3-len(st))])
    return st

File: programmers/lv1/신규아이디추천.js
function solution(new_id) {
  /*
    '/g' 일치하는 것 모두 찾기
    '^' 아닌 것 찾기
    '\w' 영어 알파벳, 숫자, 언더바
    '[^문자]' 괄호 안 문자 제외한 모든 문자 찾기
    '^문자' 특정 문자로 시작하는 것 찾기
    '문자$' 특정 문자로 끝나는 것 찾기
    '|' OR
  */
  const answer = new_id
    .toLowerCase() // 1
    .replace(/[^\w-_.]/g, '') // 2: 영단어, -, _, . 이 아닌 모든 것을 ''으로 치환
    .replace(/\.+/g, '.') // 3: .이 한 개 이상인 모든 것을 .으로 치환
    .replace(/^\.|\.$/g, '') // 4: .으로 시작하거나 .으로 끝나는 모든 것을 ''으로 치환
    .replace(/^$/, 'a') // 5: 단어가 시작끝(공백) 이면 'a' 넣기
    .slice(0, 15) // 6: 15자 초과면 15자로 맞추기
    .replace(/\.$/, ''); // 6: 끝 . 제거
  const len = answer.length;

  // 길이가 3 이상이면 그대로, 이하면 마지막 문자를 3-len 만큼 반복해 붙이기
  return len > 2 ? answer : answer + answer.charAt(len - 1).repeat(3 - len);
}
